Modules/Scripted/VectorToScalarVolume/VectorToScalarVolume.py

```diff
@@ -26,82 +63,420 @@ def __init__(self, parent):
+                                        )
+
+    self.componentsComboBox.connect('currentIndexChanged(int)', self.setComponentToExtract)
+
+  def inputVolumeNode(self):
+    return self._inputVolumeNode
+
+  def setInputVolumeNode(self, node):
+    if isinstance(node, basestring):
+      node = getNode(node)
+    self._inputVolumeNode = node
+
+  def componentToExtract(self):
+    return self._componentToExtract
+
+  def setComponentToExtract(self, value):
+    if value == self._componentToExtract:
+      return
+    self._componentToExtract = value
+    self.componentsComboBox.setCurrentIndex(value)
+
+  def conversionMethod(self):
+    return self._conversionMethod
+
+  def setConversionMethod(self, method):
+    """
+    method is data (Qt::UserRole) associated to the index in the combo box.
+    """
+    if method not in VectorToScalarVolumeLogic.CONVERSION_METHODS:
+      logging.warning("trying to set a not valid method: %s" % method)
       return
-    # check that data has enough components
-    inputImage = inputVolume.GetImageData()
-    if not inputImage or inputImage.GetNumberOfScalarComponents() < 3:
-      slicer.util.errorDisplay('Input does not have enough components for conversion',
-                               windowTitle='Vector to Scalar Volume')
+
+    if method == self._conversionMethod:
       return
-    # run the filter
-    # - extract the RGB portions
+
+    self._conversionMethod = method
+    methodIndex = self.methodSelectorComboBox.findData(method)
+    self.methodSelectorComboBox.setCurrentIndex(methodIndex)
+
+    self.componentsComboBox.enabled = (method == VectorToScalarVolumeLogic.SINGLE_COMPONENT)
+    self.componentsComboBox.clear()
+
+    if method == VectorToScalarVolumeLogic.SINGLE_COMPONENT:
+      # Check and display number of components
+      if self._inputVolumeNode is not None:
+        imageComponents = self._inputVolumeNode.GetImageData().GetNumberOfScalarComponents()
+        for comp in range(imageComponents):
+          self.componentsComboBox.insertItem(comp, str(comp))
+    else:
+      # Disable other combo box.
+      self.componentsComboBox.enabled = False # insertItem(-1, "None")
+
+
+#
+# VectorToScalarVolumeLogic
+#
+
+class VectorToScalarVolumeLogic(ScriptedLoadableModuleLogic):
+  """
+  Implement the logic to compute the transform from vector to scalar.
+  It is stateless, with the run function getting inputs and setting outputs.
+  """
+
+  LUMINANCE = 'LUMINANCE'
+  AVERAGE = 'AVERAGE'
+  SINGLE_COMPONENT = 'SINGLE_COMPONENT'
+
+  CONVERSION_METHODS = (LUMINANCE, AVERAGE, SINGLE_COMPONENT)
+
+
+  def __init__(self, parent = None):
+    ScriptedLoadableModuleLogic.__init__(self, parent)
+
+  @staticmethod
+  def isValidInputOutputData(inputVolumeNode, outputVolumeNode, conversionMethod, componentToExtract):
+    """
+    Validate parameters using the parameterNode.
```

Comment:
document what is the return value, especially when it is not trivial (not just a simple Boolean)

Reply:
Done